README.md
# PhoenixIntegration

[![Build Status](https://travis-ci.org/boydm/phoenix_integration.svg?branch=master)](https://travis-ci.org/boydm/phoenix_integration)
[![Inline docs](http://inch-ci.org/github/boydm/phoenix_integration.svg?branch=master)](http://inch-ci.org/github/boydm/phoenix_integration)
[![Module Version](https://img.shields.io/hexpm/v/phoenix_integration.svg)](https://hex.pm/packages/phoenix_integration)
[![Hex Docs](https://img.shields.io/badge/hex-docs-lightgreen.svg)](https://hexdocs.pm/phoenix_integration/)
[![Total Download](https://img.shields.io/hexpm/dt/phoenix_integration.svg)](https://hex.pm/packages/phoenix_integration)
[![License](https://img.shields.io/hexpm/l/phoenix_integration.svg)](https://github.com/boydm/phoenix_integration/blob/master/LICENSE.md)
[![Last Updated](https://img.shields.io/github/last-commit/boydm/phoenix_integration.svg)](https://github.com/boydm/phoenix_integration/commits/master)

## Overview

`PhoenixIntegration` is set of lightweight, server-side integration test functions for Phoenix.
Works within the existing `Phoenix.ConnTest` framework and emphasizes both speed and readability.

The goal is to chain together a string of requests and assertions that thoroughly
exercise your application in as lightweight and readable manner as possible.

I love the pipe `|>` command in Elixir. By using the pipe to chain together calls in an integration test, `PhoenixIntegration` is able to be very readable. Tight integration with `Phoenix.ConnTest` means the calls all use the fast-path to your application for speed.

Version 0.6 moves from Poison to Jason for Phoenix 1.4 compatibility.

Version 0.7 requires Floki 0.24.0 or higher. Otherwise it is a patch-like update.

## Documentation

You can read [the full documentation here](https://hexdocs.pm/phoenix_integration).

## Configuration

### Step 1

You need to tell phoenix_integration which endpoint to use. Add the following to your phoenix application's `config/test.exs` file.

```elixir
config :phoenix_integration,
  endpoint: MyApp.Endpoint
```

Where MyApp is the name of your application.

Do this up before compiling phoenix_integration as part of step 2. If you change the endpoint in the config file, you will need to recompile the phoenix_integration dependency.

Phoenix_integration will produce warnings if your HTML likely doesn't do what you meant. (For example, it will warn you if two text fields have the same name.) You can turn those off by adding `warnings: false` to the config.


### Step 2

Add PhoenixIntegration to the deps section of your application's `mix.exs` file

```elixir
defp deps do
  [
    # ...
    {:phoenix_integration, "~> 0.9", only: :test}
    # ...
  ]
end
```

Don't forget to run `mix deps.get`

### Step 3

Create a test/support/integration_case.ex file. Mine simply looks like this:

```elixir
defmodule MyApp.IntegrationCase do
  use ExUnit.CaseTemplate

  using do
    quote do
      use MyApp.ConnCase
      use PhoenixIntegration
    end
  end

end

```

Alternately you could place the call to `use PhoenixIntegration` in your conn_case.ex file. Just make sure it is after the definition of `@endpoint`.

### Step 4
Start writing integration tests. They should use your integration_conn.ex file. Here is a full example (just the name of the app is changed). This is from the location test/integration/page_integration_test.exs

```elixir
defmodule MyApp.AboutIntegrationTest do
  use MyApp.IntegrationCase, async: true

  test "Basic page flow", %{conn: conn} do
    # get the root index page
    get( conn, page_path(conn, :index) )
    # click/follow through the various about pages
    |> follow_link( "About Us" )
    |> follow_link( "Contact" )
    |> follow_link( "Privacy" )
    |> follow_link( "Terms of Service" )
    |> follow_button( "Accept" )
    |> follow_link( "Home" )
    |> assert_response( status: 200, path: page_path(conn, :index) )
  end

end
```

Each function in phoenix_integration accepts a conn and some other data, and returns a conn. This conn is intended to be passed into the next function via a pipe to build up a clear, readable chain of events in your test.


## Making Requests

The [`PhoenixIntegration.Requests`](https://hexdocs.pm/phoenix_integration/PhoenixIntegration.Requests.html) module contains a set functions that make requests to your application through the router.

In general, these functions look for links or forms in the html returned by a previous request. Then they make a new request to application as specified by your test. If the link wasn’t found, then an appropriate error is raised.

See [the full documentation](https://hexdocs.pm/phoenix_integration/PhoenixIntegration.Requests.html) for details.

For example, a call such as `follow_link( conn, "About Us" )`, looks in conn.body_request (which should contain html from a previous request), for an anchor tag that contains the visible text ‘About Us’. Note that it uses =~ and not == to look for the text, so you only need to specify enough text to find the link.

These functions are also pretty flexible. A call such as `follow_link( conn, "/about/us" )` recognizes that this is a path, so it looks for an anchor tag with an href equal to `“/about/us”`. Similarly, you could pass in a css-style id such as `“#about_us”` to find an anchor with the specified html id.

### Handling Redirects

All functions of the form follow_* make a request to your app. Then if a redirect is returned, makes another request following the redirect. This will go on until max_redirects is reached.

The goal is that (similar to Capybara), your integration test code looks like a set of actions that a user would actually do. To a user, redirects just happen. Clicking links and following forms are what is important.

### Submitting Forms

The `follow_form` function finds a form in the body of the previously returned conn, fills in the fields you have specified (raising an appropriate error if the form or fields aren’t found), submits the form to your application, and follows any redirects.

Used in a integration pipe chain, it looks like this:

```elixir
test "Create new user", %{conn: conn} do
  # get the root index page
  get( conn, page_path(conn, :index) )
  |> follow_link( "Sign Up" )
  |> follow_form( %{ user: %{
        name: "New User",
        email: "user@example.com",
        password: "test.password",
        confirm_password: "test.password"
      }} )
  |> assert_response(
      status: 200,
      path: page_path(conn, :index),
      html: "New User" )
end
```

The `submit_form` function is very similar, except that you handle any redirects yourself.

### Tracking multiple users

A very common scenario involves interactions between multiple users. The good news is that user state is returned in the conn from your controllers, so it is easy to track.

Is this example, I use a test_sign_in_user function (not shown), which uses token authentication so that I don’t have to pay the BCrypt price every time I run a test…

```elixir
test "admin can create a thing", %{conn: conn} do
  # create and sign in admin
  admin = test_insert_user permissions: @admin_perms
  admin_conn = test_sign_in_user(conn, admin)

  # create and sign in regular user
  user = test_insert_user
  user_conn = test_sign_in_user(conn, user)

  # admin create a new thing
  get( admin_conn, admin_path(conn, :index) )
  |> follow_link( "Create Thing" )
  |> follow_form( %{ thing: %{
        name: "New Thing"
      }} )
  |> assert_response(
      status: 200,
      path: admin_path(conn, :index),
      html: "New Thing" )

  # load the thing
  thing = Repo.get_by(Thing, name: "New Thing")
  assert thing

  # the user should be able to view the thing
  get( user_conn, page_path(conn, :index) )
  |> follow_link( thing.name )
  |> assert_response(
      status: 200,
      path: thing_path(conn, :show, thing),
      html: "New Thing"
    )
end
```

## Asserting Responses

I really wanted to see unbroken chains of piped call to make it really clear that this was a chain of events/state being tested.

The following line, which is very common in Phoenix.ConnTest controller tests works well, but doesn’t allow you to build that chain of commands.

`assert html_response(conn, 200) =~ “Some text”`

So, the `PhoenixIntegration.Assertions` module introduces two new functions, which can test multiple conditions in a single call, and always return the (unchanged) conn being tested.

See [the full documentation](https://hexdocs.pm/phoenix_integration/PhoenixIntegration.Assertions.html) for details.

I use assert_response at almost a 1:1 ratio with the various request calls, so my tests often look something like this:

```elixir
test "Basic page flow", %{conn: conn} do
# get the root index page
  get( conn, page_path(conn, :index) )
  # click/follow through the various about pages
  |> follow_link( "About Us" )
  |> assert_response( status: 200, path: about_path(conn, :index) )
  |> follow_link( "Contact" )
  |> assert_response( content_type: "text/html" )
  |> follow_link( "Privacy" )
  |> assert_response( html: "Privacy Policy" )
  |> follow_button( "Accept" )
  |> assert_response( html: "Privacy Policy" )
  |> follow_link( "Home" )
  |> assert_response( status: 200, path: page_path(conn, :index) )
end
```

To keep the chain clean and readable, each call to `assert_response` takes a conn, followed by a list of conditions to assert against. These conditions can appear multiple times in a single and will be called in the order specified.

```elixir
|> assert_response(
    status: 200,
    path:   page_path(conn, :index)
    html:   "Good Content",
    html:   "More Content"
  )
```

The `refute_response` function is very similar in form to `assert_response`, except that it refutes the given conditions. I find that it is used much less frequently, and usually prove that a response doesn’t have a specific piece of content.

```elixir
|> follow_link( "Show Thing" )
|> assert_response(
    status: 200,
    path:   thing_path(conn, :show, thing)
    html:   "Good Content"
  )
|> refute_response(
    body: "Bad Content"
  )
```

## Copyright and License

Copyright (c) 2016 Boyd Multerer

Released under the MIT License, which can be found in the repository in
[LICENSE.md](./LICENSE.md).


CHANGELOG.md
# Changelog

## v0.9.1 - 2021-06-15
  * Loosen name identification to allow form field names with characters like ? in them. Thank you @arnodirlam

## v0.9.0 - 2021-04-26
  * Fixs bug in refute_response. Thank you @StanisLove
  * update flow_assertions. Thank you @marick
  * now requires Elixir 1.10+ bumping to 0.9.0 accordingly

## v0.8.2 - 2020-04-29
  * Fixes issue #39. Nice improvements to fetch_form. Thank you @marick
  * Fix issue #41. Support phoenix 1.5 which deprecated "use Phoenix.ConnTest" in favor of "import Phoenix.ConnTest"

## v0.8.1 - 2020-02-25
  * Fixes issue #36, correctly handle other form input types. Thank you @marick for the fix.

## v0.8.0 - 2020-02-17
  * Fairly large update to handle forms with hidden fields. This update treats forms as parsed
    trees and has more informative output, and is generally more flexible. This entire update
    is brought to you by the hard work of Brian Marick (@marick on GitHub). Thank you!

## v0.7.0 - 2020-01-29
  * __Marking this as 0.7.0 because it is requires a minor version update to Floki. Otherwise
    the actual changes in phoenix_markdown are more patch-like...__
  * Changed a private functions "is_struct" to "do_is_struct" to avoid a conflict with
    the new Kernel.is_struct function in Elixir v1.10
  * Change the minimum required version of Floki to 0.24.0 and then use the new
    Floki.parse_document pattern to get rid of the deprecation warnings.
  * add .formatter.exs and format the code

## v0.6.0 - 2018-12-18
  * Moved from Poison to Jason for json parsing

## v0.5.3 - 2018-07-30
  * Merged pull request #23 from jonasschmidt to support deeply nested forms.
  * Add Travis tests for elixir 1.7

## v0.5.2 - 2018-06-11
  * Merged pull request #22 from wooga to allow single-character input names in forms

## v0.5.1 - 2018-02-01
  * fixed bug (issue #20) where it didn't find radio input fields if none were intially checked
  * removed dependency on DeepMerge

## v0.5.0 - 2018-01-16
  * added Request.click_button to find and click simple buttons on the page
  * added Request.follow_button to find, click, and follow simple buttons on the page
  * improved error message when usinga link, while asking for the wrong method.

## v0.4.1 - 2018-01-17
  * run the code through the elixir 1.6 formatter
  * update travis tests

## v0.4.0 - 2018-01-16
  * fix issue #11, was incorrectly reading the method of the form in the case of a get
  * add the fetch_form function to the Request module
  * support for nested forms. Thank you https://github.com/bitboxer
  * support follow_link for phoenix_html 2.10. Thank you https://github.com/andreapavoni
  * bump up to Elixir 1.4
  * update docs

## v0.3.0 - 2017-07-04
  * added a new :value assertion type that checks the result of a callback for truthyness
  * relaxed requirement on floki version

## v0.2.0 - 2017-01-07
  * Updated Dependencies to use version 0.13 of Floki (the html parsing enging)
  * Use updated Floki syntax when searching for links and forms by text content.
  * Errors while running a form test now display the path that was associated with the form.
    should aid in resolving issues on pages with more than one form.
    Thank you goes to https://github.com/Mbuckley0 for sorting this out.
  * Change the readme to suggest only loading phoenix_integration in test mode
  * Added support for DateTime fields in forms. Again thank you https://github.com/Mbuckley0
    for adding this feature in.

## v0.1.2 - 2016-10-03
  * Add support for file upload fields in forms

## v0.1.1 - 2016-08-13
  * Added `:assigns` option to both `Assertions.assert_response` and `Assertions.refute_response`.
  * Added `Requests.follow_fn`
  * cleaning up readme and docs

## v0.1.0 - 2016-07-28
  * First release


mix.exs
defmodule PhoenixIntegration.Mixfile do
  use Mix.Project

  @version "0.9.1"
  @source_url "https://github.com/boydm/phoenix_integration"

  def project do
    [
      app: :phoenix_integration,
      version: @version,
      elixir: "~> 1.10",
      elixirc_paths: elixirc_paths(Mix.env()),
      name: "phoenix_integration",
      deps: deps(),
      docs: docs(),
      package: package()
    ]
  end

  # Specifies which paths to compile per environment.
  defp elixirc_paths(:test), do: ["lib", "test/support"]
  defp elixirc_paths(_), do: ["lib"]

  def application do
    [applications: [:phoenix]]
  end

  defp deps do
    [
      {:phoenix, "~> 1.3"},
      {:phoenix_html, "~> 2.10"},
      {:floki, ">= 0.24.0"},
      {:jason, "~> 1.1"},
      {:flow_assertions, "~> 0.7", only: :test},
      {:ex_doc, ">= 0.0.0", only: [:dev, :docs]},
      {:inch_ex, ">= 0.0.0", only: :docs}
    ]
  end

  def docs do
    [
      extras: [
        "CHANGELOG.md": [title: "Changelog"],
        "LICENSE.md": [title: "License"],
        "README.md": [title: "Overview"]
      ],
      main: "readme",
      source_url: @source_url,
      source_ref: "v#{@version}",
      formatters: ["html"]
    ]
  end

  defp package do
    [
      description:
        "Lightweight server-side integration test functions for Phoenix." <>
          " Optimized for Elixir Pipes and the existing Phoenix.ConnTest" <>
          " framework to emphasize both speed and readability.",
      contributors: ["Boyd Multerer"],
      maintainers: ["Boyd Multerer"],
      licenses: ["MIT"],
      links: %{
        "Changelog" => "https://hexdocs.pm/phoenix_integration/changelog.html",
        "Blog Post" =>
          "https://medium.com/@boydm/integration-testing-phoenix-applications-b2a46acae9cb",
        "GitHub" => @source_url
      }
    ]
  end
end
